network/ssh.go

```diff
@@ -0,0 +1,137 @@
+/*
+   Copyright 2015 CoreOS, Inc.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package network
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/coreos/coreos-cloudinit/config"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+)
+
+const (
+	defaultUser = "core"
+	rsaKeySize  = 2048
+)
+
+// An interface for anything compatible with net.Dialer
+type Dialer interface {
+	Dial(network, address string) (net.Conn, error)
+}
+
+// SSHAgent can manage keys, updates cloud config, and loves ponies.
+// The embedded dialer is used for establishing new SSH connections.
+type SSHAgent struct {
+	agent.Agent
+	Dialer
+	User string
+}
+
+func NewSSHAgent(dialer Dialer) (*SSHAgent, error) {
+	key, err := rsa.GenerateKey(rand.Reader, rsaKeySize)
+	if err != nil {
+		return nil, err
+	}
+
+	keyring := agent.NewKeyring()
+	err = keyring.Add(key, nil, "core@default")
+	if err != nil {
+		return nil, err
+	}
+
+	return &SSHAgent{
+		Agent:  keyring,
+		Dialer: dialer,
+		User:   defaultUser,
+	}, nil
+}
+
+// Add all ssh keys to the given cloud config's default authorized_keys list.
+func (a *SSHAgent) UpdateConfig(cfg *config.CloudConfig) error {
+	keys, err := a.List()
+	if err != nil {
+		return err
+	}
+
+	for _, key := range keys {
+		cfg.SSHAuthorizedKeys = append(cfg.SSHAuthorizedKeys, key.String())
+	}
+
+	return nil
+}
+
+// Add port to host if not already set.
+func addDefaultPort(host string, port int) string {
```

Comment:
Well, meaning "set this default if a port isn't already specified in the address". Got a better name?